// @0xsequence/kit/dist/cjs/hooks/useWaasEmailAuth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useEmailAuth = useEmailAuth;
const react_1 = require("react");
const wagmiConnectors_1 = require("../connectors/wagmiConnectors");
function useEmailAuth({ connector, onSuccess }) {
    if (!connector) {
        return {
            inProgress: false,
            loading: false,
            error: undefined,
            initiateAuth: async (_email) => { },
            sendChallengeAnswer: async (_answer) => { }
        };
    }
    const [email, setEmail] = (0, react_1.useState)('');
    const [error, setError] = (0, react_1.useState)();
    const [loading, setLoading] = (0, react_1.useState)(false);
    const [instance, setInstance] = (0, react_1.useState)('');
    const [respondWithCode, setRespondWithCode] = (0, react_1.useState)();
    const getSequenceWaas = () => {
        if (!connector) {
            throw new Error('Connector is not defined');
        }
        const sequenceWaas = connector.sequenceWaas;
        if (!sequenceWaas) {
            throw new Error('Connector does not support SequenceWaaS');
        }
        return sequenceWaas;
    };
    const initiateAuth = async (email) => {
        const params = connector.params;
        const waas = getSequenceWaas();
        setLoading(true);
        setError(undefined);
        waas.onEmailAuthCodeRequired(async (respondWithCode) => {
            setRespondWithCode(() => respondWithCode);
        });
        waas
            .signIn({ email }, (0, wagmiConnectors_1.randomName)())
            .then(signInResponse => {
            onSuccess({ version: 2, signInResponse });
            if (signInResponse.email) {
                setEmail(signInResponse.email);
            }
        })
            .catch(err => {
            setError(err);
        });
        setLoading(false);
    };
    const sendChallengeAnswer = async (answer) => {
        const params = connector.params;
        const waas = getSequenceWaas();
        setLoading(true);
        setError(undefined);
        // version 2
        if (!respondWithCode) {
            throw new Error('Email v2 auth, respondWithCode is not defined');
        }
        try {
            await respondWithCode(answer);
        }
        catch (err) {
            setError(err);
        }
        finally {
            setLoading(false);
        }
    };
    const cancel = () => {
        setLoading(false);
        setRespondWithCode(null);
        setError(undefined);
    };
    return {
        inProgress: loading || !!instance,
        loading,
        error,
        initiateAuth,
        sendChallengeAnswer,
        cancel
    };
}
