// @0xsequence/kit/dist/cjs/hooks/useWaasEmailConflict.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useEmailConflict = void 0;
const waas_1 = require("@0xsequence/waas");
const react_1 = require("react");
const wagmi_1 = require("wagmi");
const accountTypeText = (info) => {
    if (!info) {
        return 'Unknown account type';
    }
    if (info.type === waas_1.IdentityType.PlayFab) {
        return 'PlayFab login';
    }
    if (info.type === waas_1.IdentityType.Email) {
        return 'Email login';
    }
    if (info.type === waas_1.IdentityType.OIDC) {
        switch (info.issuer) {
            case 'https://accounts.google.com':
                return 'Google login';
            case 'https://appleid.apple.com':
                return 'Apple login';
            default:
                return 'Unknown account type';
        }
    }
    return info.type;
};
const useEmailConflict = () => {
    const { connectors } = (0, wagmi_1.useConnect)();
    const forceCreateFuncRef = (0, react_1.useRef)(null);
    const [isOpen, toggleModal] = (0, react_1.useState)(false);
    const [emailConflictInfo, setEmailConflictInfo] = (0, react_1.useState)(null);
    const waasConnector = connectors.find(connector => !!connector.sequenceWaas);
    const waas = waasConnector?.sequenceWaas;
    (0, react_1.useEffect)(() => {
        if (waas) {
            const disposer = waas.onEmailConflict(async (info, forceCreate) => {
                forceCreateFuncRef.current = forceCreate;
                setEmailConflictInfo(info);
                toggleModal(true);
            });
            return disposer;
        }
    }, []);
    return {
        toggleEmailConflictModal: toggleModal,
        isEmailConflictOpen: isOpen,
        emailConflictInfo: {
            email: emailConflictInfo?.email ?? 'Unknown',
            type: accountTypeText(emailConflictInfo)
        },
        forceCreate: async () => {
            return forceCreateFuncRef.current?.(true);
        }
    };
};
exports.useEmailConflict = useEmailConflict;
