// @0xsequence/kit/dist/cjs/connectors/facebook/facebook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.facebook = void 0;
const wagmiConnectors_1 = require("../wagmiConnectors");
const FacebookLogo_1 = require("./FacebookLogo");
const facebook = (options) => ({
    id: 'facebook',
    isSequenceBased: true,
    logoDark: FacebookLogo_1.FacebookLogo,
    logoLight: FacebookLogo_1.FacebookLogo,
    monochromeLogoDark: (0, FacebookLogo_1.getFacebookMonochromeLogo)({ isDarkMode: true }),
    monochromeLogoLight: (0, FacebookLogo_1.getFacebookMonochromeLogo)({ isDarkMode: false }),
    // iconBackground: '#fff',
    name: 'Facebook',
    type: 'social',
    createConnector: projectAccessKey => {
        const connector = (0, wagmiConnectors_1.sequenceWallet)({
            ...options,
            connect: {
                projectAccessKey,
                ...options?.connect,
                settings: {
                    ...options?.connect?.settings,
                    signInWith: 'facebook'
                }
            }
        });
        return connector;
    }
});
exports.facebook = facebook;
