// @0xsequence/kit/dist/cjs/connectors/discord/discord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.discord = void 0;
const wagmiConnectors_1 = require("../wagmiConnectors");
const DiscordLogo_1 = require("./DiscordLogo");
const discord = (options) => ({
    id: 'discord',
    isSequenceBased: true,
    logoDark: (0, DiscordLogo_1.getDiscordLogo)({ isDarkMode: true }),
    logoLight: (0, DiscordLogo_1.getDiscordLogo)({ isDarkMode: false }),
    monochromeLogoDark: (0, DiscordLogo_1.getDiscordLogo)({ isDarkMode: true }),
    monochromeLogoLight: (0, DiscordLogo_1.getDiscordLogo)({ isDarkMode: false }),
    // iconBackground: '#fff',
    name: 'Discord',
    type: 'social',
    createConnector: projectAccessKey => {
        const connector = (0, wagmiConnectors_1.sequenceWallet)({
            ...options,
            connect: {
                projectAccessKey,
                ...options?.connect,
                settings: {
                    ...options?.connect?.settings,
                    signInWith: 'discord'
                }
            }
        });
        return connector;
    }
});
exports.discord = discord;
