// @0xsequence/kit/dist/cjs/connectors/google/google.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.google = void 0;
const wagmiConnectors_1 = require("../wagmiConnectors");
const GoogleLogo_1 = require("./GoogleLogo");
const google = (options) => ({
    id: 'google',
    isSequenceBased: true,
    logoDark: GoogleLogo_1.GoogleLogo,
    logoLight: GoogleLogo_1.GoogleLogo,
    monochromeLogoDark: (0, GoogleLogo_1.getMonochromeGoogleLogo)({ isDarkMode: true }),
    monochromeLogoLight: (0, GoogleLogo_1.getMonochromeGoogleLogo)({ isDarkMode: false }),
    name: 'Google',
    type: 'social',
    createConnector: projectAccessKey => {
        const connector = (0, wagmiConnectors_1.sequenceWallet)({
            ...options,
            connect: {
                projectAccessKey,
                ...options?.connect,
                settings: {
                    ...options?.connect?.settings,
                    signInWith: 'google'
                }
            }
        });
        return connector;
    }
});
exports.google = google;
