// @0xsequence/kit/dist/cjs/utils/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNativeTokenInfoByChainId = void 0;
const design_system_1 = require("@0xsequence/design-system");
const chains_1 = require("../chains");
const getNativeTokenInfoByChainId = (chainId, wagmiChains) => {
    // Get chain data from wagmi
    const chain = wagmiChains.find(chain => chain.id === chainId) || chains_1.chains[chainId];
    if (chain) {
        return {
            chainId: chain.id,
            name: chain.nativeCurrency.name,
            symbol: chain.nativeCurrency.symbol,
            decimals: chain.nativeCurrency.decimals,
            logoURI: (0, design_system_1.nativeTokenImageUrl)(chain.id),
            blockExplorerName: chain.blockExplorers?.default.name,
            blockExplorerUrl: chain.blockExplorers?.default.url
        };
    }
    return {
        chainId,
        name: 'Unknown',
        symbol: '???',
        decimals: 18,
        logoURI: ''
    };
};
exports.getNativeTokenInfoByChainId = getNativeTokenInfoByChainId;
