// @0xsequence/kit/dist/cjs/constants/localStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalStorageKey = void 0;
const PREFIX = '@kit';
const SETTINGS = 'settings';
const THEME = 'theme';
const ETHAUTH_PROOF = 'ethAuthProof';
const ETHAUTH_SETTINGS = 'ethAuthSettings';
const WAAS_GOOGLE_CLIENT_ID = 'waasGoogleClientId';
const WAAS_GOOGLE_ID_TOKEN = 'waasGoogleIdToken';
const WAAS_APPLE_CLIENT_ID = 'waasAppleClientId';
const WAAS_APPLE_REDIRECT_URI = 'waasAppleRedirectURI';
const WAAS_APPLE_ID_TOKEN = 'waasAppleIdToken';
const WAAS_EMAIL_ID_TOKEN = 'waasEmailIdToken';
const WAAS_ACTIVE_LOGIN_TYPE = 'waasActiveLoginType';
const WAAS_SIGN_IN_EMAIL = 'waasSignInEmail';
const SIGN_IN_EMAIL = 'signInEmail';
// TODO: remove all of this.. we should not be storing these in local storage
var LocalStorageKey;
(function (LocalStorageKey) {
    LocalStorageKey["Settings"] = "@kit.settings";
    LocalStorageKey["Theme"] = "@kit.theme";
    LocalStorageKey["EthAuthProof"] = "@kit.ethAuthProof";
    LocalStorageKey["EthAuthSettings"] = "@kit.ethAuthSettings";
    LocalStorageKey["WaasGoogleClientID"] = "@kit.waasGoogleClientId";
    LocalStorageKey["WaasGoogleIdToken"] = "@kit.waasGoogleIdToken";
    LocalStorageKey["WaasAppleClientID"] = "@kit.waasAppleClientId";
    LocalStorageKey["WaasAppleRedirectURI"] = "@kit.waasAppleRedirectURI";
    LocalStorageKey["WaasAppleIdToken"] = "@kit.waasAppleIdToken";
    LocalStorageKey["WaasActiveLoginType"] = "@kit.waasActiveLoginType";
    LocalStorageKey["WaasEmailIdToken"] = "@kit.waasEmailIdToken";
    LocalStorageKey["WaasSignInEmail"] = "@kit.waasSignInEmail";
    LocalStorageKey["SignInEmail"] = "@kit.signInEmail";
})(LocalStorageKey || (exports.LocalStorageKey = LocalStorageKey = {}));
