// @0xsequence/kit/dist/cjs/connectors/twitch/twitch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.twitch = void 0;
const wagmiConnectors_1 = require("../wagmiConnectors");
const TwitchLogo_1 = require("./TwitchLogo");
const twitch = (options) => ({
    id: 'twitch',
    isSequenceBased: true,
    logoDark: (0, TwitchLogo_1.getTwitchLogo)({}),
    logoLight: (0, TwitchLogo_1.getTwitchLogo)({}),
    monochromeLogoDark: (0, TwitchLogo_1.getTwitchLogo)({ isDarkMode: true }),
    monochromeLogoLight: (0, TwitchLogo_1.getTwitchLogo)({ isDarkMode: false }),
    // iconBackground: '#fff',
    name: 'Twitch',
    type: 'social',
    createConnector: projectAccessKey => {
        const connector = (0, wagmiConnectors_1.sequenceWallet)({
            ...options,
            connect: {
                projectAccessKey,
                ...options?.connect,
                settings: {
                    ...options?.connect?.settings,
                    signInWith: 'twitch'
                }
            }
        });
        return connector;
    }
});
exports.twitch = twitch;
