// @0xsequence/kit/dist/cjs/connectors/apple/apple.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apple = void 0;
const wagmiConnectors_1 = require("../wagmiConnectors");
const AppleLogo_1 = require("./AppleLogo");
const apple = (options) => ({
    id: 'apple',
    isSequenceBased: true,
    logoDark: (0, AppleLogo_1.getAppleLogo)({ isDarkMode: true }),
    logoLight: (0, AppleLogo_1.getAppleLogo)({ isDarkMode: false }),
    monochromeLogoDark: (0, AppleLogo_1.getAppleMonochromeLogo)({ isDarkMode: true }),
    monochromeLogoLight: (0, AppleLogo_1.getAppleMonochromeLogo)({ isDarkMode: false }),
    // iconBackground: '#fff',
    name: 'Apple',
    type: 'social',
    createConnector: projectAccessKey => {
        const connector = (0, wagmiConnectors_1.sequenceWallet)({
            ...options,
            connect: {
                projectAccessKey,
                ...options?.connect,
                settings: {
                    ...options?.connect?.settings,
                    signInWith: 'apple'
                }
            }
        });
        return connector;
    }
});
exports.apple = apple;
